broadinstitute/rawls
# [GAWB-2142] Implement metrics for workflow and submission statuses

https://broadinstitute.atlassian.net/browse/GAWB-2142

Not quite ready yet as I'm still working on testing -- including end-to-end testing with statsd and stackdriver. I thought I'd open this PR early for review/feedback though.

The following metrics were added:
- Lifetime counts of submission status, broken out by workspace
  - Answers the question:
    -  _How many submissions were submitted/aborted/etc over the last N days [for a given workspace]?_
- Lifetime counts of workflow status, broken out by (workspace, submission)
  - Answers the question:
    - _How many workflows were submitted/queued/launched/aborted/etc over the last N days [for a given workspace and submission]?_
- Current counts of workflow statuses, broken out by (workspace, submission)
  - Answers the question:
    - _How many workflows are queued/running/aborting/etc **right now** [for a given workspace and submission]?_

The last one is more of a gauge of current system status, while the first 2 are more for looking at trends over time.

Note: reviewing tests with `?w=1` is recommended as there are a lot of whitespace changes with the addition of the pattern:
```
withStatsD {
  // test code
}
```

----------

  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
